---
title: "Comparing LRT for pleiotropy v separate QTL against LOD drop in mediation analyses"
author: "Frederick Boehm"
date: "`r lubridate::now()`"
output: 
  github_document:
    toc: true
    toc_depth: 2
---

Our goal here is to examine the relationships between pleiotropy v separate QTL analyses and mediation analyses. We study with these methods the chromosome 2 hotspot in the Keller et al 2018 pancreatic islet expression data.


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
```

First, we load the Keller 2018 data.

```{r}
load("../data-to-ignore/Attie_DO378_eQTL_viewer_v1.Rdata")
readRDS("../data/hotspot_expr_tib2_keller_chr2.rds") -> hotspot_expr
readRDS("../data/keller2018-chr2-local-expr.rds") -> local_expr
```

We get the annotation information on the genes that map to chromosome 2's hotspot.

```{r}
dataset.islet.rnaseq$annots %>%
  filter(hotspot == "chr2")
```


```{r}
# what is the ENSEMBL id for Hnf4a?
hnf4a_id <- "ENSMUSG00000017950"
# define mediator - a numeric vector with mouse ids as names
med <- local_expr %>%
  select(hnf4a_id)
rownames(med) <- local_expr$mouse_id
# define target - a numeric vector with names (mouse ids)
tar_id <- "ENSMUSG00000000544"
tar <- hotspot_expr %>%
  select(tar_id) %>%
  unlist()
names(tar) <- hotspot_expr$mouse_id
```

We still need to define the driver for input to `intermediate::mediation_scan`.

```{r}
# define driver
## to define driver, we need to find the peak position for Hnf4a gene expression levels
library(qtl2)
# convert to needed phenotypes input format for scan1
as.matrix(local_expr[, 1:21]) -> local_expr_mat
rownames(local_expr_mat) <- local_expr$mouse_id %>% unlist() %>% as.character()
#
as.matrix(hotspot_expr[, 1:139]) -> hotspot_expr_mat
rownames(hotspot_expr_mat) <- hotspot_expr$mouse_id %>% unlist() %>% as.character()
#
# isolate additive covariates
dataset.islet.rnaseq$covar -> addcovar
```




We save the scan1 outputs, because they take some time to re-run.

```{r}
s1_local <- readRDS("../data/s1_local.rds")
s1_hotspot <- readRDS("../data/s1_hotspot.rds")
```



We now search the outputs of `scan1` to identify peaks:

```{r}
find_peaks(s1_local, map = map, threshold = 20) -> local_peaks
find_peaks(s1_hotspot, map = map) -> hotspot_peaks
```

```{r}
map_chr2_tib <- tibble(pos = map$`2`, index = 1:length(map$`2`))
hnf4a_peak_position <- local_peaks %>%
  filter(lodcolumn == hnf4a_id) %>% 
  left_join(map_chr2_tib, by = "pos") %>%
  select(index) %>% 
  unlist() %>%
  as.numeric()
```



```{r}
dri <- genoprobs[[2]][rownames(genoprobs[[2]]) %in% hotspot_expr$mouse_id, , hnf4a_peak_position] 
# We take the second item in the genoprobs list because the peaks are on Chromosome TWO
```

We also need to define an annotation dataframe.

```{r}
ann <- dataset.islet.rnaseq$annots %>%
  filter(gene_id == hnf4a_id) %>% 
  rename(pos = middle, id = gene_id)
```

Let's subset the driver and the covariates:

```{r}
addcovar2 <- addcovar[rownames(addcovar) %in% names(tar), ]
dri2 <- dri[rownames(dri) %in% names(tar), ]
```




```{r}
rna_annot <- dataset.islet.rnaseq$annots %>%
  as_tibble() %>%
  select(gene_id, symbol)
```


## Mediation analyses with `intermediate`

```{r}
library(intermediate)
```



```{r}
# before setting up dri2, we need to find peak indices for the hotspot traits. 
# Join with map object
h2 <- hotspot_peaks %>%
  filter(chr == 2) %>%
  left_join(map_chr2_tib)
# set up drivers
#dri <- genoprobs[[2]][rownames(genoprobs[[2]]) %in% hotspot_expr$mouse_id, , h2$index] 

# mediation scan calls
out <- list()
for (i in 1:ncol(hotspot_expr_mat)){
  #med_lod_drop[i] <- mediation_scan(hotspot_expr_mat[ , i], driver = dri[ , , i], mediator = med, annotation = ann, covar = addcovar2)$lod
  out[[i]] <- fit1_med(hotspot_expr_mat[ , i], driver = dri, mediator = med, annotation = ann, covar = addcovar2)
}
med_lod_drop <- purrr::map_dbl(out, function(x)x$lod)
target_fit <- purrr::map_dbl(out, function(x)attributes(x)$targetFit)
(meds <- tibble(local_gene_id = hnf4a_id, nonlocal_gene_id = colnames(hotspot_expr)[-140], med_lod_drop, target_fit))
```







## Tabulate lod drops

```{r}
sum(meds$med_lod_drop > 1.5)
sum(meds$med_lod_drop <= 1.5)
```

## Mediation analyses without `intermediate`

First, we define functions.

```{r}
calc_loglik_lm <- function(residuals){
  sum(residuals ^ 2) -> rss
  ll <- - length(residuals) * log(rss) / 2
  return(ll)
}

fit1_lm <- function(driver,
                     target,
                     mediator,
                     addcovar = NULL
){
  ll_d <- calc_loglik_lm(lm(target ~ -1 + driver + addcovar)$residuals)
  ll_dm <- calc_loglik_lm(lm(target ~ -1 + driver + mediator + addcovar)$residuals)
  ll_1 <- calc_loglik_lm(lm(target ~ -1 + addcovar)$residuals)
  ll_1m <- calc_loglik_lm(lm(target ~ -1 + mediator + addcovar)$residuals)
  out <- tibble::tibble(ll_d, ll_dm, ll_1, ll_1m)
  return(out)
}


fit1_ll <- function(driver,
                     target,
                     mediator,
                     addcovar = NULL
){
  ll_d <- logLik(lm(target ~ -1 + driver + addcovar))
  ll_dm <- logLik(lm(target ~ -1 + driver + mediator + addcovar))
  ll_1 <- logLik(lm(target ~ -1 + addcovar))
  ll_1m <- logLik(lm(target ~ -1 + mediator + addcovar))
  out <- tibble::tibble(ll_d, ll_dm, ll_1, ll_1m)
  return(out)
}

fit1_med_qtl2 <- function(driver,
                     target,
                     mediator,
                     addcovar = NULL,
                     kinship = NULL
){
  lod_no_med <- qtl2::fit1(genoprobs = driver, 
                           pheno = target, 
                           kinship = kinship, 
                           addcovar = addcovar
                           )$lod
  lod_med <- qtl2::fit1(genoprobs = driver, 
                        pheno = target, 
                        kinship = kinship, 
                        addcovar = cbind(addcovar, mediator)
                        )$lod
  
  out <- tibble::tibble(lod_no_med, lod_med)
  return(out)
}
```

Now, let's use the two functions, without using kinship objects, to get "proportion LOD drop" and "LOD drop".

```{r}
fit1_ll(driver = dri,
         target = hotspot_expr_mat[ , 1], 
         mediator = unlist(med), 
         addcovar = addcovar2) %>%
  mutate(lod_without_med = (ll_d - ll_1) / log(10), lod_with_med = (ll_dm - ll_1m) / log(10), lod_drop = lod_without_med - lod_with_med, lod_drop_proportion = lod_drop / lod_without_med)
```
```{r}
fit1_lm(driver = dri,
         target = hotspot_expr_mat[ , 1], 
         mediator = unlist(med), 
         addcovar = addcovar2) %>%
  mutate(lod_without_med = (ll_d - ll_1) / log(10), lod_with_med = (ll_dm - ll_1m) / log(10), lod_drop = lod_without_med - lod_with_med, lod_drop_proportion = lod_drop / lod_without_med)
```




```{r}
fit1_med_qtl2(driver = dri,
         target = hotspot_expr_mat[ , 1], 
         mediator = unlist(med), 
         addcovar = addcovar2) %>%
  mutate(lod_drop = lod_no_med - lod_med, lod_drop_proportion = lod_drop / lod_no_med)
```

I see that there's a slight discrepancy between the two results. I'm concerned that the function that uses `logLik`, ie, the function `fit1_med` is doing something, and I can't quite tell what exactly it does... since I don't know how exactly to see the code for `stats::logLik.lm`.

I then tried to see if the function `fitDefault` from `intermediate` gives the same answers as my `fit1_med_qtl2`. The answer is yes.

Now, let's do the regressions for each column in `hotspot_expr_mat`. 

```{r}
out <- list()
for (i in 1:139){
  out[[i]] <- fit1_ll(driver = dri,
         target = hotspot_expr_mat[ , i], 
         mediator = unlist(med), 
         addcovar = addcovar2)
}
med_hnf4a_out <- out %>%
  bind_rows() %>%
  mutate(lod_without_med = (ll_d - ll_1) / log(10), lod_with_med = (ll_dm - ll_1m) / log(10), lod_drop = lod_without_med - lod_with_med, lod_drop_proportion = lod_drop / lod_without_med, nonlocal_gene_id = colnames(hotspot_expr_mat), local_gene_id = hnf4a_id)
```

## Set up mediation analyses for all 21 local genes

First, we want to set up the `dri21` driver object. To do that, we need to know the peak positions for each local trait.

```{r}
indices <- local_peaks %>%
  left_join(map_chr2_tib) %>%
  select(index) %>%
  unlist()
```

```{r}
dri21 <- dri <- genoprobs[[2]][rownames(genoprobs[[2]]) %in% hotspot_expr$mouse_id, , indices]
cbind(dimnames(dri21)[[3]], 
indices)
```
```{r}
med21 <- local_expr_mat[ , colnames(local_expr_mat) %in% local_peaks$lodcolumn]
med21 %>% colnames()
```





```{r}
out21 <- list()
foo <- list()
for (j in 1:21){
  for (i in 1:139){
   foo[[i]]<- fit1_ll(driver = dri[ , , j],
         target = hotspot_expr_mat[ , i], 
         mediator = med21[ , j], 
         addcovar = addcovar2)
  }
  out21[[j]] <- foo
}
```

```{r}
med_out <- out21 %>%
  purrr::map(bind_rows)
m2_out <- med_out %>%
  bind_rows() %>%
  as_tibble() %>%
  mutate(local_gene_id = rep(colnames(med21), each = 139), nonlocal_gene_id = rep(colnames(hotspot_expr_mat), times = 21)) %>%
  mutate(lod_without_med = (ll_d - ll_1) / log(10), lod_with_med = (ll_dm - ll_1m) / log(10), lod_drop = lod_without_med - lod_with_med, lod_drop_proportion = lod_drop / lod_without_med)

```





## Tabulating the pleiotropy v separate QTL LRT statistics

Our goal is to analyze the results of pvl scan results files from CHTC.

```{r}
library(qtl2chtc)
```

```{r}
dir(path = "../results/pvl-run2001/", pattern = "*.txt") -> fns
  lrt <- numeric()
  i <- 1
  for (fn in fns){
    read.table(file.path("../results/pvl-run2001/", fn)) -> scan_out
    qtl2pleio::calc_lrt_tib(scan_out) -> lrt[i]
    i <- i + 1
  }
tibble::tibble(fns, lrt) -> r2001
```


Let's parse the file names in `r2001` into meaningful fields, like gene names.

```{r}
library(stringr)
```

```{r}
splitted <- r2001$fns %>% 
  str_split(pattern = "_") 

my_split <- function(string_vec, pattern = ".txt"){
  foo <- str_split(string_vec, pattern)
  sapply(FUN = function(x)x[1], X = foo)
}
res2001 <- sapply(FUN = function(x)x[1:4], X = splitted) %>% 
  t() %>%
  as_tibble() %>%
  rename(run_num = V1, job_num = V2, local_gene_id = V3) %>%
  mutate(nonlocal_gene_id = my_split(V4)) %>%
  select(- V4) %>%
  mutate(lrt = r2001$lrt) %>%
  arrange(desc(lrt)) %>%
  left_join(rna_annot, by = c("nonlocal_gene_id" = "gene_id")) %>%
  rename(nonlocal_gene_symbol = symbol) %>%
  left_join(rna_annot, by = c("local_gene_id" = "gene_id")) %>%
  rename(local_gene_symbol = symbol)
  

  
```


```{r}
hist(res2001$lrt)
```

```{r}
sum(res2001$lrt > 4)
```



## Join pleiotropy v separate QTL LRT tibble with LOD drops tibble

```{r}
library(plotly)
```
    
```{r}
pp <- med_hnf4a_out %>% 
  left_join(res2001) %>%
  ggplot() + aes(x = lrt, y = lod_drop, colour = nonlocal_gene_symbol) + geom_point() + theme(legend.position="none")
ggplotly(pp, tooltip = c("colour", "x", "y")) 
```


```{r}
pp <- med_hnf4a_out %>% 
  left_join(res2001) %>%
  ggplot() + aes(x = lrt, y = lod_drop_proportion, colour = nonlocal_gene_symbol) + geom_point() + theme(legend.position="none")
ggplotly(pp, tooltip = c("colour", "x", "y")) 

```

## Work with all 2919 test statistics & mediation analyses

```{r}
pp2 <- m2_out %>% 
  left_join(res) %>%
  ggplot() + aes(x = lrt, y = lod_drop_proportion, colour = local_gene_symbol) + geom_point() + theme(legend.position="none")
ggplotly(pp2, tooltip = c("colour", "x", "y")) 
  
```


```{r}
pp2 <- m2_out %>% 
  left_join(res) %>%
  filter(local_gene_symbol %in% c("Hnf4a", "Ptgis")) %>%
  ggplot() + aes(x = lrt, y = lod_drop_proportion, colour = nonlocal_gene_symbol) + geom_point() + theme(legend.position="none") + facet_wrap(~ local_gene_symbol, ncol = 1)
ggplotly(pp2, tooltip = c("colour", "x", "y")) 
```
Note that Gpr20 has a high LRT for pleiotropy v separate QTL with Hnf4a, yet it has a small LRT when paired with Ptgis. In both cases, the proportion drop in LOD is very small. 

Let's look at the table of results that involve Gpr20

```{r}
m2_out %>% 
  left_join(res) %>%
  filter(nonlocal_gene_symbol == "Gpr20") %>%
  arrange(desc(lod_drop_proportion))
```
It looks like Gpr20 has small LOD scores for most loci examined here.

```{r}
gpr20_out <- m2_out %>% 
  left_join(res) %>%
  filter(nonlocal_gene_symbol == "Gpr20") %>%
  arrange(desc(lod_without_med))
```

```{r}
local2 <- local_peaks %>%
  as_tibble() %>%
  rename(local_gene_lod = lod, local_gene_id = lodcolumn)
local2 %>%
  left_join(gpr20_out) %>%
  tidyr::gather(lod_drop_proportion, lrt, key = "method", value = "test_stat") %>%
  ggplot() + geom_point(aes(x = pos, y = test_stat, colour = method)) + facet_wrap(~method, ncol = 1, scales="free") 
# add vertical line for nonlocal gene's univariate QTL peak?   
  
  
```

```{r}
library(ggrepel)
```



```{r}
for (i in 1:139){
  bar <- m2_out %>% 
    left_join(res) %>%
    left_join(local2) %>%
    filter(nonlocal_gene_id == colnames(hotspot_expr_mat)[i]) %>%
    tidyr::gather(lod_drop, lrt, key = "method", value = "test_stat")
    ggplot(bar) + geom_point(aes(x = pos, y = test_stat, colour = method)) + geom_label_repel(aes(x = pos, y = test_stat, label = local_gene_symbol),
                  box.padding   = 0.35, 
                  point.padding = 0.5,
                  segment.color = 'grey50') + facet_wrap(~ method, scales = "free", ncol = 1) 
    ggsave(file.path("figures", paste0(bar$nonlocal_gene_symbol[1] , "-pvl-med.png")))
  }  
```

