---
title: "Visualizing allele effects patterns across traits that map to chromosome 2 hotspot"
author: "Frederick Boehm"
date: "9/10/2018"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
```


## Goal

Goal in this Rmd file is to write code needed for creation of a heatmap of allele effects values for every trait that maps to chromosome 2 hotspot and many local traits near the hotspot.

## Resources for heatmaps in R

Here are links to some potentially valuable resources for heatmap creation in R:



## Karl's suggestions for clusterings of traits

In our meeting last week (September 6 2018), Karl suggested several ways to cluster or organize the traits within the heatmap. These include:

- ordering based on LOD peak position   
- ordering by LOD peak height   
- ordering by size of effect for a single founder allele
- order by one or more visual features in the allele effects plots.

To fully realize these methods, I need to create the univariate allele effects plots. I'll add a section below in which I do just that. Below, I already have code that calculates the allele effects for each trait. Thus, writing code to plot those effects (ie, on chr2 only) should be straightforward with qtl2 functions.



## Plan

First, we need to scan chromosome 2 with each trait. From the scan outputs, we can calculate the allele effects at the position of the LOD peak. Once we have the 8 founder allele effects per trait, we can create a heatmap

## Univariate scans of chromosome 2

First, we need to read the .Rdata file that we downloaded from DataDryad.

```{r}
load("../data-to-ignore/Attie_DO378_eQTL_viewer_v1.Rdata")
readRDS("../data/hotspot_expr_tib2_keller_chr2.rds") -> hotspot_expr
readRDS("../data/keller2018-chr2-local-expr.rds") -> local_expr21
readRDS("../data/keller2018-chr2-local-expr-114.rds") -> local_expr114
local_expr <- local_expr114 %>%
  full_join(local_expr21)
```

We also loaded the RDS files that contain the expression values for the traits of interest.

The genotypes probabilities are their own object (loaded with the DataDryad file), `genoprobs`.

```{r}
library(qtl2)
# convert to needed phenotypes input format for scan1
as.matrix(local_expr[, c(1:114, 116)]) -> local_expr_mat 
# col 115 is the mouse_id
rownames(local_expr_mat) <- local_expr[, 115] %>% unlist() %>% as.character()
#
as.matrix(hotspot_expr[, 1:139]) -> hotspot_expr_mat
rownames(hotspot_expr_mat) <- hotspot_expr[, 140] %>% unlist() %>% as.character()
#
# isolate additive covariates
dataset.islet.rnaseq$covar -> addcovar
```



```{r}
s1_local <- readRDS("../data/s1_local.rds")
s1_hotspot <- readRDS("../data/s1_hotspot.rds")
```

We now search the outputs of `scan1` to identify peaks:

```{r}
find_peaks(s1_local, map = map, threshold = 20) -> local_peaks
find_peaks(s1_hotspot, map = map) -> hotspot_peaks
```

Note above that we use the default value of `threshold` when calling `find_peaks` for the hotspot traits. This ensures that we consider all 169 traits. we'll subset on chromosome 2 peaks below.

Next, we calculate allele effects for each trait on chromosome 2. We keep separate the local and hotspot traits.

```{r}
# local traits
local_effects <- list()
for (i in 1:115){
  scan1coef(genoprobs = genoprobs[ , 2], 
            pheno = local_expr_mat[ , i],
            kinship = K[[2]],
            addcovar = addcovar,
            reml = TRUE
            ) -> local_effects[[i]]
}    
# hotspot traits
hotspot_effects <- list()
for (i in 1:139){
  scan1coef(genoprobs = genoprobs[ , 2], 
            pheno = hotspot_expr_mat[ , i],
            kinship = K[[2]],
            addcovar = addcovar,
            reml = TRUE
            ) -> hotspot_effects[[i]]
}    
```

Now, we want to isolate, for each trait, the 8-long vector of founder allele effects at the trait's corresponding peak position.

We need to use the R object `map` (from the Datadryad file).

```{r}
map[[2]]
```

Note that in our output from the calls to `find_peaks`, we have position in Mb. However, in the outputs of the calls to `scan1coef`, we have no positions - we do, however, know that the marker ordering is maintained.

So, we need to identify the index - on chromosome 2 - that corresponds to each trait's peak position.

```{r}
library(tidyverse)
# convert map for chr2 to a tibble
m2 <- map[[2]] %>% 
  as_tibble() %>%
  mutate(index = 1:4709)
local_peaks2 <- local_peaks %>%
  as_tibble() %>%
  left_join(m2, by = c("pos" = "value"))
hotspot_peaks2 <- hotspot_peaks %>%
  as_tibble() %>%
  filter(chr == "2") %>%
  left_join(m2, by = c("pos" = "value"))
```

We define a helper function:

```{r}
get_effects <- function(marker_index, allele_effects_matrix, map){
  marker_name <- names(map[marker_index])
  allele_effects_matrix[rownames(allele_effects_matrix) == marker_name, 1:8]
}
```

```{r}
hotspot_effects_at_peaks <- list()
for (i in 1:nrow(hotspot_peaks2)){
  hotspot_effects_at_peaks[[i]] <- get_effects(hotspot_peaks2$index[i],
                                               allele_effects_matrix = hotspot_effects[[i]], 
                                               map = map[[2]]
                                               ) # need map for chromosome 2 here.
}
local_effects_at_peaks <- list()
for (i in 1:nrow(local_peaks2)){
  local_effects_at_peaks[[i]] <- get_effects(local_peaks2$index[i], 
                                             allele_effects_matrix = local_effects[[i]],
                                             map = map[[2]]
                                             ) # need map for chromosome 2 here.
}
```

We then arrange the two sets of allele effects into matrices.

```{r}
# organize effects into a matrix
local_effects_at_peaks_mat <- sapply(FUN = function(x)x, 
                                     X = local_effects_at_peaks) %>% t()
rownames(local_effects_at_peaks_mat) <- local_peaks2$lodcolumn
hotspot_effects_at_peaks_mat <- sapply(FUN = function(x)x, 
                                       X = hotspot_effects_at_peaks) %>% t()
rownames(hotspot_effects_at_peaks_mat) <- hotspot_peaks2$lodcolumn
```

```{r}
saveRDS(local_effects_at_peaks_mat, "../data/founder_allele_effects_local_traits_matrix.rds")

saveRDS(hotspot_effects_at_peaks_mat, "../data/founder_allele_effects_hotspot_traits_matrix.rds")

saveRDS(hotspot_effects, "../data/founder_allele_effects_hotspot.rds")
saveRDS(local_effects, "../data/founder_allele_effects_local.rds")
```



## Code to create allele effects plots for chromosome 2 for each univariate trait.

Our goal here is to present the univariate allele effects plots for each trait being studied in this document - ie, 21 local traits and 139 nonlocal traits.

```{r, eval = FALSE}
# first, subset s1_local to get only those traits that are on chromosome 2
# we have 4711 markers on chr 1 and 4709 on chr 2. Thus chr2 starts with marker 4710 and ends with 9420


for (i in 1:ncol(local_expr_mat)){
  plot_coefCC(x = local_effects[[i]], 
              map = map, scan1_output = NULL, add = FALSE
              ) # be careful - do we need to subset the scan1_output to only those markers on chromosome 2?
} # I want to add gene transcript names to each plot.
# certainly, we need to identify the Hnf4a gene, but should get ids for others, too!

```








We first prepare a helper function:

```{r}
tidy_effects <- function(effects_mat){
  effect_size <- as.vector(effects_mat)
  founder_allele <- rep(colnames(effects_mat), each = nrow(effects_mat))
  gene_expr_id <- rep(rownames(effects_mat), times = ncol(effects_mat))
  return(tibble(gene_expr_id, founder_allele, effect_size))
}
```



## Clustering genes' allele effects

```{r}
local_effects_at_peaks_mat %>%
  t() %>%
  cor() %>%
  is.na() %>%
  sum()
local_effects_at_peaks_mat %>%
  #t() %>%
  dist() %>%
  hclust() %>%
  plot()
local_hc <- local_effects_at_peaks_mat %>%
  dist() %>%
  hclust()
```

## Exploring with `d3heatmap` R package

I decided that I want to try using the `d3heatmap` package. It's on CRAN.

```{r}
library(d3heatmap)
```

```{r}
d3heatmap(local_effects_at_peaks_mat, Colv = FALSE
          )
```

I can't remember the Ensembl id for Hnf4a. Let's get it:

```{r}
dataset.islet.rnaseq$annots %>% filter(symbol == "Hnf4a")
```

We see that, in the heatmap above, *Hnf4a* is located in the first row.

Honestly, the other "local" genes seem to look rather distinct from Hnf4a in this plot.

Now, let's look at a heatmap of the nonlocal traits:

```{r}
d3heatmap(hotspot_effects_at_peaks_mat, Colv = FALSE
          )

```

Notice the similarities in the nonlocal traits in the heatmap above and the top row (Hnf4a) of the heatmap of the local traits. Wow! This seems very exciting - it's like nearly all of the nonlocal traits share some heatmap similarities with the heatmap row that is Hnf4a.

Can we create a heatmap of the nonlocal traits, with ordering of rows determined not by dendrogram but by univariate LOD score for mapping to the hotspot?

It seems that, to create this heat map with `d3heatmap`, we actually need to reorder the allele effects matrix before inputting it to `d3heatmap`. In our call to `d3heatmap`, of course, we need to tell it to NOT reorder the rows.

```{r}
ensembl_ids_by_lod <- hotspot_peaks2 %>%
  arrange(desc(lod)) %>%
  select(lodcolumn) %>%
  unlist()
```


```{r}
ord_lod <- match(table = rownames(hotspot_effects_at_peaks_mat), x = ensembl_ids_by_lod)
hotspot_effects_ordered_by_lod <- hotspot_effects_at_peaks_mat[ord_lod,]
ensembl_ids_by_lod
(rownames(hotspot_effects_ordered_by_lod) == ensembl_ids_by_lod) %>%
  sum()
```

Now that we have ordered the input matrix, let's call the heatmap function:

```{r}
d3heatmap(hotspot_effects_ordered_by_lod, Rowv = FALSE, Colv = FALSE)
```

## Correlations among allele effects vectors

```{r}

```







### Mediation test results in the heatmaps

We now want to see if we can find the mediation test results for these hotspot traits with Hnf4a. If we can find them, then we can also order the input matrix (ie, that of allele effects values for the 139 hotspot traits) by, say, their lod score drop when conditioning on Hnf4a expression levels.





